ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java

```diff
@@ -0,0 +1,1760 @@
+      }
+    }
+    // reload index to make sure journal is on only the latest log segment
+    state.reloadIndex(true, false);
+    // compact with deleteReferenceTimeMs < the earliest delete time. None of the deleted keys count as deleted.
+    compactAndVerify(expiryTimeAndSegmentsUnderCompaction.getSecond(), deleteReferenceTimeMs, true);
+    // ensure that idsToExamine can still be fetched
+    state.reloadIndex(true, false);
+    for (MockId id : idsToExamine) {
+      // should not throw exception since they should be untouched.
+      // Data has already been verified if this is true (by the verifiers).
+      state.index.getBlobReadInfo(id, EnumSet.of(StoreGetOptions.Store_Include_Deleted)).close();
+    }
+  }
+
+  /**
+   * Tests the case where deletes and expired blobs are interspersed and the expired blobs are eligible for cleanup
+   * but deleted blobs (also includes blobs that have been put and deleted in the same index segment).
+   * @throws Exception
+   */
+  @Test
+  public void interspersedDeletedAndExpiredBlobsTest() throws Exception {
+    refreshState(false, false);
+    state.properties.put("store.index.max.number.of.inmem.elements", Integer.toString(5));
+    state.initIndex(new MetricRegistry());
+
+    int numFinalSegmentsCount = 3;
+    long expiryTimeMs = getInvalidationTime(numFinalSegmentsCount + 1);
+    // fill up one segment.
+    writeDataToMeetRequiredSegmentCount(1, null);
+
+    // 1. Put entry that contains a delete entry in the same index segment and is not counted as deleted.
+    // won't be cleaned up.
+    IndexEntry entry = state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time).get(0);
+    MockId delUnexpPutSameIdxSegId = (MockId) entry.getKey();
+    String logSegmentName = entry.getValue().getOffset().getName();
+    state.addDeleteEntry(delUnexpPutSameIdxSegId);
+
+    // 2. Put entry that has expired and contains a delete entry in the same index segment. Does not count as deleted
+    // but is expired.
+    // will be cleaned up, but delete record remains
+    MockId delExpPutSameIdxSegId =
+        (MockId) state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, 0).get(0).getKey();
+    state.addDeleteEntry(delExpPutSameIdxSegId);
+
+    // 3. Put entry that has expired.
+    // will be cleaned up.
+    state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, 0).get(0).getKey();
+
+    // 4. Put entry that will expire (but is valid right now).
+    // won't be cleaned up.
+    MockId willExpPut =
+        (MockId) state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, expiryTimeMs).get(0).getKey();
+
+    // 5. Put entry that will be deleted.
+    // won't be cleaned up.
+    MockId willBeDelPut =
+        (MockId) state.addPutEntries(1, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time).get(0).getKey();
+
+    // rollover
+    // 6. Put entry with an expiry time  that is not expired and has a delete entry in the same index segment and is not
```

Comment:
Because "store.index.max.number.of.inmem.elements" is set in the beginning of the function and based on that (and the previous entries added in the same function), it is possible to determine where the entry will be. Does that answer your question or could you think of cases where this would not hold true ? (Other than code/logic changes in which case the tests need changing anyway).